# Lab 7: 1D Composite Simpson's Rule/Gaussian Quadrature

### Consider the integral

$$
\int_{0.75}^{2.3} \sin^2 x - 2x \sin x + 1 \, dx \tag{1}
$$

(a) Use your Python function `compositeSimpson` to approximate the value of the integral (1) with a number of subintervals $n = 2, 4, 8, 16, 32$. Notice that when $n = 2$, the Composite Simpson’s rule is the same as Simpson’s rule.

(b) Use your Python function `gaussianQuad` to approximate the value of the integral (1) with a number of points used in the quadrature $n = 2, 3, 4, 5$.

(c) Given that the antiderivative of $\sin^2 x - 2x \sin x + 1$ is $\frac{3}{2} x - \frac{1}{4} \sin 2x - 2 \sin x + 2x \cos x$. Find the absolute error of the approximations you found in (a) and (b).

(d) Consider your results with integral (1). In view of the number of points used in the methods, which method, Composite Simpson’s rule or Gaussian quadrature, is more economical? That is, which method uses fewer points in the method and has more accurate approximations?

(e) For the errors of the approximations you found in (a), assume the error is of order $p$; that is, assume $\text{err}(h) = Ch^p$. Rewrite the relation into $\ln(\text{err}(h)) = \ln(C) + p \ln(h)$. Now collect the absolute errors we should have

$$
\begin{bmatrix}
1 & \ln(h_1) \\
1 & \ln(h_2) \\
\vdots & \vdots \\
1 & \ln(h_k) \\
\end{bmatrix}
\begin{bmatrix}
\ln C \\
p \\
\end{bmatrix}
=
\begin{bmatrix}
\ln(\text{err}(h_1)) \\
\ln(\text{err}(h_2)) \\
\vdots \\
\ln(\text{err}(h_k)) \\
\end{bmatrix}.
$$

Let the matrix be $A$ and the right-hand side vector be $\vec{b}$. Use the NumPy command `numpy.linalg.lstsq(A, b)[0]` to find $p$. Is the result consistent with the theoretical order of the error for the Composite Simpson’s method?


In [1]:
# a
from NumericalMethodsCode.compositeSimpson import compositeSimpson
import numpy as np
f = lambda x: np.sin(x)**2 - 2*x*np.sin(x) + 1
I1 = compositeSimpson(f, 0.75, 2.3, 2)
print('n=2: I1 =',I1)
I2 = compositeSimpson(f, 0.75, 2.3, 4)
print('n=4: I2 =',I2)
I3 = compositeSimpson(f, 0.75, 2.3, 8)
print('n=8: I3 =',I3)
I4 = compositeSimpson(f, 0.75, 2.3, 16)
print('n=16: I4 =',I4)
I5 = compositeSimpson(f, 0.75, 2.3, 32)
print('n=32: I5 =',I5)

n=2: I1 = -1.4537931762047078
n=4: I2 = -1.4671468479293726
n=8: I3 = -1.4677056220507707
n=16: I4 = -1.4677373438196153
n=32: I5 = -1.4677392796675288


In [2]:
# b
from NumericalMethodsCode.gaussianQuad import gaussianQuad
J1 = gaussianQuad(f, 0.75, 2.3, 2)
print('n=2: J1 =', J1)
J2 = gaussianQuad(f, 0.75, 2.3, 3)
print('n=3: J2 =', J2)
J3 = gaussianQuad(f, 0.75, 2.3, 4)
print('n=4: J3 =', J3)
J4 = gaussianQuad(f, 0.75, 2.3, 5)
print('n=5: J4 =', J4)

n=2: J1 = -1.4772971426576487
n=3: J2 = -1.4674499819353886
n=4: J3 = -1.4677428298057102
n=5: J4 = -1.467739384257614


In [3]:
# c
# finding the actual value
p = lambda x: (3/2)*x - (1/4)*np.sin(2*x) - 2*np.sin(x) + 2*x*np.cos(x)
p_b = p(2.3)
p_a = p(0.75)
p_ = p_b - p_a

print('Absolute Errors:')
print("composite Simpson's rule")
abserr1 = abs(p_ - I1)
print('n=2:', abserr1)
abserr2 = abs(p_ - I2)
print('n=4:', abserr2)
abserr3 = abs(p_ - I3)
print('n=8:', abserr3)
abserr4 = abs(p_ - I4)
print('n=16:', abserr4)
abserr5 = abs(p_ - I5)
print('n=32:', abserr5)
print("")
print("Gaussian Quadrature")
abserr01 = abs(p_ - J1)
print('n=2:', abserr01)
abserr02 = abs(p_ - J2)
print('n=3:', abserr02)
abserr03 = abs(p_ - J3)
print('n=4:', abserr03)
abserr04 = abs(p_ - J4)
print('n=5:', abserr04)

Absolute Errors:
composite Simpson's rule
n=2: 0.013946231740606052
n=4: 0.0005925600159413236
n=8: 3.378589454317016e-05
n=16: 2.064125698586494e-06
n=32: 1.2827778506085963e-07

Gaussian Quadrature
n=2: 0.009557734712334787
n=3: 0.00028942600992531986
n=4: 3.4218603963687144e-06
n=5: 2.3687699801655526e-08


c)

Gaussian quadrature is more economical since it requires fewer points in the method, and it also contains the lowest absolute errors, meaning it has more accurate approximations.

In [9]:
# d
import numpy as np
b = 2.3
a = 0.75
h1 = (b-a)/2
h2 = (b-a)/4
h3 = (b-a)/8
h4 = (b-a)/16
h5 = (b-a)/32

A = np.array([[1, np.log(h1)],
             [1, np.log(h2)],
             [1, np.log(h3)],
             [1, np.log(h4)],
             [1, np.log(h5)]])
b = np.array([[np.log(abserr1)],
             [np.log(abserr2)],
             [np.log(abserr3)],
             [np.log(abserr4)],
             [np.log(abserr5)]])

np.linalg.lstsq(A,b)[0]

C:\Users\ntlmp\AppData\Local\Temp\ipykernel_32408\2639518365.py:22: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  np.linalg.lstsq(A,b)[0]


array([[-3.36022205],
       [ 4.16257756]])

The result is conistent with the theoretical order of the error for the Composite Simpson's method since it is $O(h^4)$, and the output for p is

```
array([[-3.36022205],
       [ 4.16257756]])
```
So the error is of order(4.16257756).